### ISSUE TITLE
dataclass: Type of "<parameter>" could not be determined because it refers to itself

### ISSUE THREAD

**Issue by ealap** [microsoft/pyright#7516]
**Describe the bug**
Shown below are two similar codes that both works at runtime. The issue only occurs if I put the type `Path` or if I change the type of `dir_b` to other type while still referencing the value of `dir_a`
![image](https://github.com/microsoft/pyright/assets/15620712/0fa2752e-421d-4ce3-b6e9-e073f6f758ff)

This produces the error:
```python
@dataclass()
class SampleClassA:
    name: str = "ealap/sample"
    dir_a: Path = Path.home().joinpath(f"source/{name}")
    dir_b: Path = dir_a.joinpath("path/to/b")
```

These are OK for pyright:
```python
@dataclass()
class SampleClassA:
    name: str = "ealap/sample"
    dir_a = Path.home().joinpath(f"source/{name}")
    dir_b = dir_a.joinpath("path/to/b")


@dataclass()
class SampleClassB:
    name: str = "ealap/sample"
    dir_a: Path = Path.home().joinpath(f"source/{name}")
    dir_b: str = dir_a.joinpath("path/to/b").stem


class SampleClassC:
    name: str = "ealap/sample"
    dir_a: Path = Path.home().joinpath(f"source/{name}")
    dir_b: Path = dir_a.joinpath("path/to/b")
```

**VS Code extension or command-line**
pyright cli: v1.1.354
vscode: Pylance v2024.3.1

**Comment by Azureblade3808** [comment]
![image](https://github.com/microsoft/pyright/assets/15389744/b5846c5d-cd7a-4184-9298-879f1cca2233)

Failed to repro. Does this issue exist in a fresh project?

**Comment by erictraut** [comment]
I'm able to repro the problem, but it's intermittent. Very strange. I'll investigate further.

**Comment by ealap** [comment]
Indeed, it is intermittent. I just opened the affected project now after a machine restart and the issue is not shown anymore.

**Comment by erictraut** [comment]
I now understand what's happening, but I don't yet have a good solution.

Pyright is a just-in-time type evaluator. It can evaluate the type of any expression. If that type depends on the types of other symbols, it recursively evaluates those types until it has the information it needs.

In this case, it is being asked to evaluate the type of `dir_b` on the last line of the code sample. The type of this symbol depends on the type of `dir_a`. This symbol depends on the expression `Path.home...`, but it also depends on the type of the class that contains it. Enums and dataclasses require special processing, so the type checker needs to understand whether `SampleClassA` is a dataclass or an enum. To determine this, it needs to determine the type of `SampleClassA`. For normal (non-dataclass) classes, this would simply involve evaluating the base class list in the `class` statement. However, dataclasses require additional processing because they involve several synthesized class-scoped symbols including an `__init__` method. This `__init__` method is synthesized based on the dataclass members and their types, which includes `dir_a` and `dir_b`. Evaluating the types of these symbols results in a circularity, which is what you see reported.

The reason you're hitting this is because you're doing something very unusual in this class definition. You're using the value of a class variable `dir_a` in the class body to calculate the value of another class variable `dir_b`. Both of these class variables are later converted into instance variables by the `dataclass` decorator.

Here's a way you can avoid the circularity:

```python
from dataclasses import dataclass
from pathlib import Path

_name_default = "ealap/sample"
_dir_a_default = Path.home().joinpath(f"source/{_name_default}")
_dir_b_default = _dir_a_default.joinpath("path/to/b")

@dataclass()
class SampleClassA:
    name: str = _name_default
    dir_a: Path = _dir_a_default
    dir_b: Path = _dir_b_default
```

I need to think more about whether there's a way I can break this circularity in the type evaluator. I don't currently see any straightforward way to do it. If I'm unable to find a way, then I will likely close this issue as "won't fix".

**Comment by Azureblade3808** [comment]
So it seems to be the same issue I've met by using a pattern in my data-class defining code, which I almost neglected.

I like to define value converters as private static functions just before their corresponding fields, so that related code could stay close. For example --

```python
from datetime import date

import attrs

@attrs.frozen
class Period:
    @staticmethod
    def __convert_start(date_like: date | int | str, /) -> date: ...

    start: date = attrs.field(converter=__convert_start)

    @staticmethod
    def __convert_end(date_like: date | int | str, /) -> date: ...

    end: date = attrs.field(converter=__convert_end)
```

Type errors may appear when code gets modified in my project, but always disappear with a re-saving of the file that raises the errors.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.358, which I just published.
